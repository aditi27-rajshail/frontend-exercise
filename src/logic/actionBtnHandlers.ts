export const actionBtnHandlers = {
  selectExchangeBlockchain: () => {
    console.log("取引所 / ブロックチェーンを選択しましょう");
    // Implement the action logic here
  },
  reflectTradeHistory: () => {
    console.log("取引履歴を反映させましょう");
    // Implement the action logic here
  },
  eliminateRequiredIdentificationDeFi: () => {
    console.log("DeFi取引における「要識別」を解消しましょう");
    // Implement the action logic here
  },
  eliminateUnclassifiedTxn: () => {
    console.log("未分類取引を解消しましょう");
    // Implement the action logic here
  },
  // Add more handlers for other actions
};
