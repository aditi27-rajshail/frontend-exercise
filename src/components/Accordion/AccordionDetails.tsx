/**
 * AccordionDetails component for displaying detailed content in an accordion-style container.
 * @param {AccordionDetailsData} accordionDetails - The data for the accordion details.
 * @param {Function} onToggle - The function to handle the toggle action.
 * @param {boolean} isChecked - A flag to indicate whether the item is checked.
 * @param {boolean} disabled - A flag to indicate whether the accordion details are disabled.
 * @returns {React.Element} - The rendered AccordionDetails component.
 */

import Button from "../Button";
import { checkCircle, circleDashed } from "../../assets";
import Alert from "../Alert";
import { AccordionDetailsData } from "../../types";
import { actionBtnHandlers } from "../../logic/actionBtnHandlers";

interface AccordionDetailsProps {
  accordionDetails: AccordionDetailsData;
  onToggle: () => void;
  isChecked: boolean;
  disabled: boolean;
}

const AccordionDetails = ({
  accordionDetails,
  onToggle,
  isChecked,
  disabled,
}: AccordionDetailsProps) => {
  const callActionHandler = (actionKey: keyof typeof actionBtnHandlers) => {
    const handler = actionBtnHandlers[actionKey];
    if (handler) {
      handler();
    } else {
      console.log(`No action handler for key: ${actionKey}`);
    }
  };
  return (
    <div className="border border-solid border-neutral500 mt-4 p-4 rounded-lg">
      <div className="md:flex lg:gap-[77px] md:gap-[38px]">
        <div className="flex items-start">
          <img
            src={isChecked ? checkCircle : circleDashed}
            alt={isChecked ? "check circle" : "circle dashed"}
            width={24}
            height={24}
            className={` mr-[8px] ${!disabled && "cursor-pointer easeout"}`}
            onClick={!disabled ? onToggle : undefined}
          />
          {/* Content */}
          <div>
            <div className={`${disabled && "disabled"}`}>
              <h4 className="accordion-detail-title mb-2">
                {accordionDetails.title}
              </h4>
              {accordionDetails?.tag && (
                <span className="chip text-info600 h-[20px] py-0 px-2">
                  {accordionDetails.tag}
                </span>
              )}

              <p className="accordion-detail-sub-text mt-2">
                {accordionDetails.subText}
              </p>
            </div>
            {accordionDetails?.info && (
              <Alert text={accordionDetails.info} severity="info" />
            )}
          </div>
        </div>
        <div className="flex flex-col items-center justify-center mt-3 md:mt-0">
          <Button
            buttonText={accordionDetails.buttonText}
            disabled={disabled}
            onClick={() => callActionHandler(accordionDetails.btnAction)}
          />
        </div>
      </div>
    </div>
  );
};

export default AccordionDetails;
