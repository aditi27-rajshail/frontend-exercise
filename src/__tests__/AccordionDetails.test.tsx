import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import AccordionDetails from "../components/Accordion/AccordionDetails";
import { mainData } from "../constants";
import { actionBtnHandlers } from "../logic/actionBtnHandlers";

describe("Accordion Details Component", () => {
  describe("Accordion Details renderes component correctly", () => {
    beforeEach(() => {
      render(
        <AccordionDetails
          accordionDetails={mainData[0].accordionDetails[0]}
          onToggle={() => {}}
          isChecked={false}
          disabled={false}
        />
      );
    });

    it("renders the title", () => {
      const title = mainData[0].accordionDetails[0].title;
      const titleElement = screen.getByText(title);
      expect(titleElement.textContent).toBe(title);
    });

    it("renders the subText", () => {
      const subText = mainData[0].accordionDetails[0].subText;
      const subTextElement = screen.getByText(subText);
      expect(subTextElement.textContent).toBe(subText);
    });

    it("renders the empty dashed icon if isChecked is false", () => {
      const checkboxImage = screen.getByRole("img", {
        name: /circle dashed/i,
      });
      expect(checkboxImage).toBeDefined();
    });

    it("renders the green check icon if isChecked is true", () => {
      render(
        <AccordionDetails
          accordionDetails={mainData[1].accordionDetails[0]}
          onToggle={() => {}}
          isChecked={true}
          disabled={false}
        />
      );
      const checkboxImage = screen.getByRole("img", {
        name: /check circle/i,
      });
      expect(checkboxImage).toBeDefined();
    });

    it("renders the chip span if tag is provided", () => {
      const tag = mainData[0].accordionDetails[0].tag;
      const subTextElement = screen.getByText(tag);
      expect(subTextElement.textContent).toBe(tag);
    });

    it("renders the Button Component with text and right arrow", () => {
      const buttonElement = screen.getByRole("button", {
        name: `${mainData[0].accordionDetails[0].buttonText} Right Arrow`,
      });
      expect(buttonElement).toBeDefined();
    });

    it("calls button click handler when the button is clicked", () => {
      const actionKey = mainData[0].accordionDetails[0].btnAction;
      const actionHandlerSpy = vi.spyOn(actionBtnHandlers, actionKey);
      const buttonElement = screen.getByRole("button", {
        name: `${mainData[0].accordionDetails[0].buttonText} Right Arrow`,
      });
      fireEvent.click(buttonElement);

      expect(actionHandlerSpy).toHaveBeenCalled();
    });

    it("renders the Alert Component if info is provided", () => {
      const infoText = mainData[1].accordionDetails[0].info;
      render(
        <AccordionDetails
          accordionDetails={mainData[1].accordionDetails[0]}
          onToggle={() => {}}
          isChecked={false}
          disabled={false}
        />
      );
      const alertElement = screen.getByText(infoText);
      expect(alertElement.textContent).toBe(infoText);
    });
  });

  it("calls onToggle when the check icon is clicked", () => {
    const onToggleMock = vi.fn();
    render(
      <AccordionDetails
        accordionDetails={mainData[0].accordionDetails[0]}
        onToggle={onToggleMock}
        isChecked={false}
        disabled={false}
      />
    );
    const checkboxImage = screen.getByRole("img", {
      name: /circle dashed/i,
    });
    fireEvent.click(checkboxImage);
    expect(onToggleMock).toHaveBeenCalledTimes(1);
  });

  it("does not call onToggle when disabled is true", () => {
    const onToggleMock = vi.fn();
    render(
      <AccordionDetails
        accordionDetails={mainData[0].accordionDetails[0]}
        onToggle={onToggleMock}
        isChecked={false}
        disabled={true}
      />
    );
    const checkboxImage = screen.getByRole("img", {
      name: /circle dashed/i,
    });
    fireEvent.click(checkboxImage);
    expect(onToggleMock).not.toHaveBeenCalled();
  });

  it("matches the snapshot", () => {
    const { container } = render(
      <AccordionDetails
        accordionDetails={mainData[0].accordionDetails[0]}
        onToggle={() => {}}
        isChecked={false}
        disabled={false}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
