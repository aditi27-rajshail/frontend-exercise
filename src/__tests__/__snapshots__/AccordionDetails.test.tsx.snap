// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion Details Component > matches the snapshot 1`] = `
<div>
  <div
    class="border border-solid border-neutral500 mt-4 p-4 rounded-lg"
  >
    <div
      class="md:flex lg:gap-[77px] md:gap-[38px]"
    >
      <div
        class="flex items-start"
      >
        <img
          alt="circle dashed"
          class=" mr-[8px] cursor-pointer easeout"
          height="24"
          src="/src/assets/circle-dashed.svg"
          width="24"
        />
        <div>
          <div
            class="false"
          >
            <h4
              class="accordion-detail-title mb-2"
            >
              取引所 / ブロックチェーンを選択しましょう
            </h4>
            <span
              class="chip text-info600 h-[20px] py-0 px-2"
            >
              13 取引所/ブロックチェーン選択済み
            </span>
            <p
              class="accordion-detail-sub-text mt-2"
            >
              取引をしたことのある取引所 / ブロックチェーンをすべて選択してください。対応していない取引所や取引所外での取引経験がある場合は「カスタム」を選択してください。
            </p>
          </div>
        </div>
      </div>
      <div
        class="flex flex-col items-center justify-center mt-3 md:mt-0"
      >
        <button
          class="false easeout flex h-[32px] w-[104px] p-[12px] items-center justify-center bg-primary500 rounded-[4px]"
        >
          <p
            class="btn-text mr-2"
          >
             
            選択する
          </p>
          <img
            alt="Right Arrow"
            height="16"
            src="/src/assets/arrow-right-bold.svg"
            width="16"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
